drivers: modem: receiver: uart fixes and switch to ring buffer

GeorgeCGV opened the pull request:
The modem receiver doesn't work stable enough with `k_pipe`. I experienced data loss, and rx freezes.

After some investigation, it seems like the receiver omits almost every note regarding kernel and uart API usage. Therefore I propose the following modifications:

- Switched to ring buffer from pipe API
> Kernel doesn't allow for an ISR to send or receive data to/from
> a pipe even when there is no attempt to wait for space/data.
- Corrected uart fifo usage, sending is done with `uart_poll_out`
> Likewise, not calling this function from an ISR if uart_irq_tx_ready() returns true may lead to undefined behavior, e.g. infinite interrupt loops.
- Bug: `rx_sem` shall be given back when the buffer|pipe was full and uart data has been flushed|drained
- Correct error code, use `ENODEV `when `device_get_binding` fails
- Added comments and asserts
_To make it easier for others_

GeorgeCGV commented:
@mike-scott

mike-scott commented:
@GeorgeCGV First, let me thank you for taking a look at the modem receiver code.  I haven't had time to go back and clean it up for some time.

I think there is too much going on for just 1 commit in this PR.  Perhaps we can split this up to increase the readability and understand the changes:
- "uart fixes" here is moving away from the interrupt based fifo_fill API and using poll_out?  I know fifo_fill is buggy and the logic needs to be cleaned up, but poll_out is terribly slow and I would hate to have that as the basis for all modems using the receiver..  I think we should look at the correct usage of fifo_fill and fix the logic here.
- Agree that the documentation needs love and I appreciate the addition here.  Can we have that in a separate commit?
- Switching from pipe to ring buffer should be in a separate commit
- Not sure on the movement on mdm_receiver_setup() and mdm_receiver_context_from_id().. seems arbitrary?
- Changing error # returned (and adding logging) in mdm_receiver_register() should be in a separate commit

mike-scott reviewed (changes requested):
@GeorgeCGV First, let me thank you for taking a look at the modem receiver code.  I haven't had time to go back and clean it up for some time.

I think there is too much going on for just 1 commit in this PR.  Perhaps we can split this up to increase the readability and understand the changes:
- "uart fixes" here is moving away from the interrupt based fifo_fill API and using poll_out?  I know fifo_fill is buggy and the logic needs to be cleaned up, but poll_out is terribly slow and I would hate to have that as the basis for all modems using the receiver..  I think we should look at the correct usage of fifo_fill and fix the logic here.
- Agree that the documentation needs love and I appreciate the addition here.  Can we have that in a separate commit?
- Switching from pipe to ring buffer should be in a separate commit
- Not sure on the movement on mdm_receiver_setup() and mdm_receiver_context_from_id().. seems arbitrary?
- Changing error # returned (and adding logging) in mdm_receiver_register() should be in a separate commit

GeorgeCGV commented:
> * "uart fixes" here is moving away from the interrupt based fifo_fill API and using poll_out?  I know fifo_fill is buggy and the logic needs to be cleaned up, but poll_out is terribly slow and I would hate to have that as the basis for all modems using the receiver..  I think we should look at the correct usage of fifo_fill and fix the logic here.

I couldn't find any good example of correct fifo_fill usage, except one which involves 2 separate buffers - rx and tx being processed in one isr. In my experience Hayes command set tend to be quite synchronious and I am not sure about performance gain from another buffer in tx. But this can be tried out.

> * Not sure on the movement on mdm_receiver_setup() and mdm_receiver_context_from_id().. seems arbitrary?

This separates internal (_static_) functions and the public ones. 

> * Agree that the documentation needs love and I appreciate the addition here.  Can we have that in a separate commit?
> * Switching from pipe to ring buffer should be in a separate commit
> * Changing error # returned (and adding logging) in mdm_receiver_register() should be in a separate commit

Can be done.

mike-scott commented:
> I couldn't find any good example of correct fifo_fill usage, except one which involves 2 separate buffers - rx and tx being processed in one ISR. In my experience, Hayes command set tends to be quite synchronous and I am not sure about performance gain from another buffer in tx, as it will put additional limits to manage on the tx side and possibly extra data copying. Maybe this can be added extra as another option for tx communication.

Let's put these changes in a commit of it's own to be debated.  I'm inclined to move back just because the modem receiver will then work across most SoC.  We can investigate how to use the interrupt-based APIs correctly after LTS release.

> This separates internal (_static_) functions and the public ones into different file sections.

Ok, let's have this movement in a separate commit.

Thanks!

Commit:
drivers: modem: receiver: correct device bind error type

Log possible uart device binding error and change
the error type to a more correct one.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: modem: receiver: switch to ring buffer from pipe api

Replaces used k_pipe with ring_buffer API which is safe for ISR.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: modem: receiver: rearrange and comment

Rearranges functions to public and private groups,
and adds missing function comments.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: modem: receiver: fix sending

Send implementation shall return 0 for ok,
or < 0 for an error.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: modem: receiver: use asserts internally

Switch to asserts where applicable for internal
functions.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: modem: receiver: replace fifo_fill with poll_out

Repalces send fifo_fill implementation with poll_out,
this makes the receiver usefull on most SoC.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: modem: receiver: guards against 0 size

Protects against 0 size in send, receive, and
registration.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

mike-scott commented on drivers/modem/modem_receiver.c:
```diff
@@ -119,7 +116,9 @@ int mdm_receiver_recv(struct mdm_receiver_context *ctx,
-	return k_pipe_get(&ctx->uart_pipe, buf, size, bytes_read, 1, K_NO_WAIT);
+	*bytes_read = ring_buf_get(&ctx->rx_rb, buf, size);
+
+	return 0;
```
This is a documentation issue, but should be:
return (*bytes_read == 0 ? -EIO : 0);

Otherwise, we'll endless loop here waiting for < 0 result (meaning buffer is empty)
See: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/modem/wncm14a2a.c#L1081

mike-scott replied:
Of course I could change WNCM14A2A code to check for 0 bytes read and that might solve this issue as well.

GeorgeCGV replied:
great idea

GeorgeCGV commented on drivers/modem/modem_receiver.c:
```diff
@@ -98,15 +95,15 @@ static void mdm_receiver_isr(struct device *uart_dev)
 		rx = uart_fifo_read(ctx->uart_dev, read_buf, sizeof(read_buf));
 		if (rx > 0) {
-			ret = k_pipe_put(&ctx->uart_pipe, read_buf, rx,
-					 &bytes_written, rx, K_NO_WAIT);
```
yeap, missed that

GeorgeCGV commented:
> But why consider zero bytes as error? I mean that if the caller of `mdm_receiver_recv` called with size of zero, then my intuition expects that the return value would be zero.

> Yes, I noted originally that I should probably change how this function is used in the Wistron modem so that we can return 0 bytes read w/o error condition.
> 
> I'll setup a patch for that change.

@gon1332, I don't think there is a correct answer.  Reading with the size of 0 can be considered as an invalid operation or an invalid argument - as the operation makes no sense, when you try to read something you shall specify what your expectations or limits are. Sometimes reading data and checking data availability is a separate set of functions (e.g. `peek`).

@gon1332, @mike-scott,  I would suggest to have a look at existing API across the system to see what is the most common way being used. Then depending on this come to a conclusion, so that API stays consistent across as many places as possible.

mike-scott commented on drivers/modem/modem_receiver.c:
```diff
@@ -119,7 +115,9 @@ int mdm_receiver_recv(struct mdm_receiver_context *ctx,
-	return k_pipe_get(&ctx->uart_pipe, buf, size, bytes_read, 1, K_NO_WAIT);
+	*bytes_read = ring_buf_get(&ctx->rx_rb, buf, size);
+
+	return (*bytes_read == 0 ? -EIO : 0);
```
Let's put this back to just returning 0 here (sorry for the noise).  I'll put up the PR for Wistron modem fix in a bit.   bytes_read == 0 isn't really an error.

GeorgeCGV replied:
Well... let it be so

mike-scott commented:
Added @nashif as this does fix some inconsistencies across several SoCs where the interrupt-based UART wasn't working the same way.  Not sure if it's a bug per-say and belongs in LTS.  But for those working on future modem drivers this is a really good fix IMHO.

@nashif use your best judgement on whether this should be merged for v1.14

mike-scott commented:
@nashif @galak The more I think about it, the more I'd like this in 1.14.

Reasoning:
1. It's LTS which means it will "live" a bit longer than most Zephyr releases
2. There are 2-3 new modem drivers being developed across several different SoCs.  I'm assuming the development for those will continue in the LTS branch.  There's going to be confusion as to why the drivers work correctly on some SoCs and not others w/o this PR.
3. The danger level for merging this is very low.  It's a small subsystem which I (for the most part) maintain.

Drivers that I know of which are in-progress:
- [me] Particle.io Boron LTE support (ublox sara r41xx series)
- [Andy Gross] I forgot to write down the modem he's working on
- [Geo] Not sure I can disclose modem model, but he also has a WIP driver

GeorgeCGV commented:
Additionally, pull request #13386 prepares kconfig for additions. I hope it can be merged as well.
